add substring search to smith-waterman and new tests

gspowley opened the pull request:
#1629 @akiezun @droazen @lbergelson
Added a substring search to `SWPairwiseAlignment.align `to avoid running the full Smith-Waterman when the query is found in the reference without any indels. The performance benefit of this code will be data dependent. In the current HaplotypeCaller test, >80% of the Smith-Waterman calls are filtered by the substring search.

Added tests to cover all of the overhang strategies.

**Note:** The substring search only works for the `SOFTCLIP `and `IGNORE `overhang strategies. The `INDEL `and `LEADING_INDEL `can result in more complicated CIGAR strings. See the `SWPairwiseAlignmentUnitTest.testSubstringMatchIndelLong` and `SWPairwiseAlignmentUnitTest.testSubstringMatchLeadingIndelLong` tests for examples.

droazen commented:
For @akiezun to review. Recommend that we merge this after the `HaplotypeCaller` branch rather than before to make it easier to evaluate whether the output changes at all.

akiezun commented:
With some tweaks I see a good improvement. 

First i tried it on our testing bam (on dr_runnable_haplotypecaller) and saw some speedup:

before, in GVCF mode
```
real	6m6.430s
user	7m54.014s
sys	0m4.540s
```

after
```
real	5m58.246s
user	7m53.737s
sys	0m4.403s
```

Then I rewrote it a bit by avoiding the `new String` calls and directly using the static `lastIndexOf` method from String that works on arrays (required copying code from the Java libraries because the right method is not accessible). The result was a 30s improvement over baseline (6%)
```
real	5m44.213s
user	7m26.240s
sys	0m3.916s
```

So that would be a nice improvement but I'm not sure if copy/pasting code from the JDK is kosher license-wise. To be completely safe, @gspowley can you implement a clean-room version of `lastIndexOf(byte[] source, byte[] target)` so that we can replace `new String(reference).lastIndexOf(new String(alternate))` with `lastIndexOf(reference, alternate)`? Or let's find a library that we can use for this. I want to avoid `String` creation here because it's pretty expensive.

droazen commented:
@akiezun Can you please test with `--activeRegionsOnly` and `-ERC GVCF` on `dr_runnable_haplotypecaller`, and see if the gains hold up (or increase/decrease)?

akiezun commented:
I dont think i see this option. maybe you need to push your current code

droazen commented:
@akiezun It's there in the branch, try it: `--activeRegionsOnly`

droazen commented:
(if you really don't see it, you may need to bring your copy of my branch up-to-date via: `git fetch && git reset --hard d910985883ae1197a5a1bf0adf80d44ee9031928`

gspowley commented:
@akiezun I had a clean-room version of `lastIndexOf(byte[] source, byte[] target)` ready to go. It's based on KMP algorithm described in [Algorithms, 4th edition](http://algs4.cs.princeton.edu/home/), modified to work on `byte []` and to search from right to left.

Here are some profiling results from runs with instrumentation code.
- `Elapsed time` is reported by HaplotypeCaller
- `Smith-Waterman search` is time spent in lastIndexOf
- `Smith-Waterman full` is time spent in full Smith-Waterman

**Command**
```
./gatk-launch HaplotypeCaller -R src/test/resources/large/human_g1k_v37.20.21.fasta -I src/test/resources/large/CEUTrio.HiSeq.WGS.b37.NA12878.20.21.bam -O out.g.vcf -ERC GVCF
```
**Original Code**
```
Elapsed time: 5.35 minutes.
Time in Smith-Waterman search = 0.004659962
Time in Smith-Waterman full   = 44.342853681
Smith-Waterman calls = 103995 (0 substring match)
```
**Code with KMP search**
```
Elapsed time: 4.90 minutes.
Time in Smith-Waterman search = 0.129977206
Time in Smith-Waterman full   = 27.955518434000002
Smith-Waterman calls = 103995 (85728 substring match)
```
**Code with String.lastIndexOf**
```
Elapsed time: 4.92 minutes.
Time in Smith-Waterman search = 0.081916556
Time in Smith-Waterman full   = 27.025490193000003
Smith-Waterman calls = 103995 (85728 substring match)
```

There wasn't a lot of difference in KMP vs. String, so I didn't include KMP in the first commit. But, these numbers may not capture the `String` creation penalty.

akiezun commented:
thanks @gspowley I'll measure it again tomorrow. I see the KMP code has an array allocation - I wonder if this accounts for the relatively small difference between string and KMP. I'll also try again with my mud-room implementation and compare. Also, my Mac is clearly slower than what you're running on :) I'll try David's suggestion too and limit to active regions.

droazen commented:
Yes, please both of you try to do your profiling with `--activeRegionsOnly` and `-ERC GVCF`, since I expect this to become the default mode shortly, and the profile is VERY different.

gspowley commented:
Good point @akiezun on the array allocation in KMP. Here's a clean room brute force `lastIndexOf`. Initial testing show identical output from HaplotypeCaller and better performance than KMP.

I'll push this later today and include the Smith-Waterman profiling code.

``` java
    private int lastIndexOf(final byte[] reference, final byte[] query) {
        int N = reference.length;
        int M = query.length;

        for (int i = N - M; i >= 0; i--) {
            int j = 0;
            while (j < M && reference[i+j] == query[j]) {
                j++;
            }
            if (j == M) {
                return i;
            }
        }

        return -1;
    }

```

akiezun commented:
super! thanks @gspowley

gspowley commented:
@akiezun @droazen Here are my performance results for the latest commit using `-ERC GVCF` and `--activeRegionsOnly`.

**Original Code**
```
Elapsed time: 2.54 minutes.
Time in Smith-Waterman search = 0.0037741310000000004 sec
Time in Smith-Waterman full   = 19.315050559 sec
Total Smith-Waterman calls    = 84948 (0 substring matches = 0.00 %)
```

**Code with lastIndexOf for array**
```
Elapsed time: 2.43 minutes.
Time in Smith-Waterman search = 0.020596938000000002 sec
Time in Smith-Waterman full   = 12.461222691000001 sec
Total Smith-Waterman calls    = 84948 (69338 substring matches = 81.62 %)
```

~5% improvement for this test case.

The Smith-Waterman profiling is currently enabled by default. There is also code to dump every Smith-Waterman call and result (currently disabled by default).

akiezun commented:
I like it. I think it should go in. Can you remove the profiling code, squash and merge?  Do move lastInstanceOf to Utils though and add some tests.

gspowley commented:
@akiezun Moved `lastInstanceOf` to `Utils`, removed profiling code, added tests, and squashed/merged. This can go in after `HaplotypeCaller` as we discussed with @droazen.

akiezun commented:
👍  merge right after HC goes in to save @droazen the headaches of rebasing

sputnikci commented on src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java:
```diff
@@ -532,4 +532,66 @@ public void testEqualRange(byte[] arr1, int from, int to, byte[] arr2, boolean e
+        final String reference = "AAAA";
+        final String query     = "AAAAAAA";
+
+        final int result = Utils.lastIndexOf(reference.getBytes(), query.getBytes());
```
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.UtilsUnitTest\.testLastIndexOfQueryTooLong\(\): String\.getBytes\(\)

sputnikci replied:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.UtilsUnitTest\.testLastIndexOfQueryTooLong\(\): String\.getBytes\(\)

sputnikci commented on src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java:
```diff
@@ -532,4 +532,66 @@ public void testEqualRange(byte[] arr1, int from, int to, byte[] arr2, boolean e
+
+        // match right boundary of reference
+        String query = "TGGGG";
+        int result = Utils.lastIndexOf(reference.getBytes(), query.getBytes());
```
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.UtilsUnitTest\.testLastIndexOfLastBoundaries\(\): String\.getBytes\(\)

sputnikci replied:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.UtilsUnitTest\.testLastIndexOfLastBoundaries\(\): String\.getBytes\(\)

sputnikci commented on src/test/java/org/broadinstitute/hellbender/utils/UtilsUnitTest.java:
```diff
@@ -532,4 +532,66 @@ public void testEqualRange(byte[] arr1, int from, int to, byte[] arr2, boolean e
+            }
+            
+            final int result = Utils.lastIndexOf(reference, query);
+            final int expected = new String(reference).lastIndexOf(new String(query));
```
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.UtilsUnitTest\.testLastIndexOfRandom\(\): new String\(byte\[\]\)

sputnikci replied:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.UtilsUnitTest\.testLastIndexOfRandom\(\): new String\(byte\[\]\)

sputnikci commented on src/test/java/org/broadinstitute/hellbender/utils/smithwaterman/SWPairwiseAlignmentUnitTest.java:
```diff
@@ -117,4 +117,119 @@ public void  testForIdenticalAlignmentsWithDifferingFlankLengths() {
+        final String read      = match;
+        final int expectedStart = 3;
+        final String expectedCigar = "5M";
+        final SWPairwiseAlignment sw = new SWPairwiseAlignment(reference.getBytes(), read.getBytes(),
```
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchSoftclip\(\): String\.getBytes\(\)

sputnikci replied:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchSoftclip\(\): String\.getBytes\(\)

sputnikci commented on src/test/java/org/broadinstitute/hellbender/utils/smithwaterman/SWPairwiseAlignmentUnitTest.java:
```diff
@@ -117,4 +117,119 @@ public void  testForIdenticalAlignmentsWithDifferingFlankLengths() {
+        final String read      = match;
+        final int expectedStart = 0;
+        final String expectedCigar = "3D5M";
+        final SWPairwiseAlignment sw = new SWPairwiseAlignment(reference.getBytes(), read.getBytes(),
```
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchIndel\(\): String\.getBytes\(\)

sputnikci replied:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchIndel\(\): String\.getBytes\(\)

sputnikci commented on src/test/java/org/broadinstitute/hellbender/utils/smithwaterman/SWPairwiseAlignmentUnitTest.java:
```diff
@@ -117,4 +117,119 @@ public void  testForIdenticalAlignmentsWithDifferingFlankLengths() {
+        final String read      = match;
+        final int expectedStart = 0;
+        final String expectedCigar = "3D5M";
+        final SWPairwiseAlignment sw = new SWPairwiseAlignment(reference.getBytes(), read.getBytes(),
```
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchLeadingIndel\(\): String\.getBytes\(\)

sputnikci replied:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchLeadingIndel\(\): String\.getBytes\(\)

sputnikci commented on src/test/java/org/broadinstitute/hellbender/utils/smithwaterman/SWPairwiseAlignmentUnitTest.java:
```diff
@@ -117,4 +117,119 @@ public void  testForIdenticalAlignmentsWithDifferingFlankLengths() {
+        final String read      = match;
+        final int expectedStart = 3;
+        final String expectedCigar = "5M";
+        final SWPairwiseAlignment sw = new SWPairwiseAlignment(reference.getBytes(), read.getBytes(),
```
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchIgnore\(\): String\.getBytes\(\)

sputnikci replied:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchIgnore\(\): String\.getBytes\(\)

sputnikci commented on src/test/java/org/broadinstitute/hellbender/utils/smithwaterman/SWPairwiseAlignmentUnitTest.java:
```diff
@@ -117,4 +117,119 @@ public void  testForIdenticalAlignmentsWithDifferingFlankLengths() {
+        final String read      = "AAAAAAA";
+        final int expectedStart = 359;
+        final String expectedCigar = "7M";
+        final SWPairwiseAlignment sw = new SWPairwiseAlignment(reference.getBytes(), read.getBytes(),
```
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchSoftclipLong\(\): String\.getBytes\(\)

sputnikci replied:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchSoftclipLong\(\): String\.getBytes\(\)

sputnikci commented on src/test/java/org/broadinstitute/hellbender/utils/smithwaterman/SWPairwiseAlignmentUnitTest.java:
```diff
@@ -117,4 +117,119 @@ public void  testForIdenticalAlignmentsWithDifferingFlankLengths() {
+        final String read      = "AAAAAAA";
+        final int expectedStart = 0;
+        final String expectedCigar = "1M358D6M29D";
+        final SWPairwiseAlignment sw = new SWPairwiseAlignment(reference.getBytes(), read.getBytes(),
```
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchIndelLong\(\): String\.getBytes\(\)

sputnikci replied:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchIndelLong\(\): String\.getBytes\(\)

sputnikci commented on src/test/java/org/broadinstitute/hellbender/utils/smithwaterman/SWPairwiseAlignmentUnitTest.java:
```diff
@@ -117,4 +117,119 @@ public void  testForIdenticalAlignmentsWithDifferingFlankLengths() {
+        final String read      = "AAAAAAA";
+        final int expectedStart = 0;
+        final String expectedCigar = "1M1D6M";
+        final SWPairwiseAlignment sw = new SWPairwiseAlignment(reference.getBytes(), read.getBytes(),
```
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchLeadingIndelLong\(\): String\.getBytes\(\)

sputnikci replied:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchLeadingIndelLong\(\): String\.getBytes\(\)

sputnikci commented on src/test/java/org/broadinstitute/hellbender/utils/smithwaterman/SWPairwiseAlignmentUnitTest.java:
```diff
@@ -117,4 +117,119 @@ public void  testForIdenticalAlignmentsWithDifferingFlankLengths() {
+        final String read      = "AAAAAAA";
+        final int expectedStart = 359;
+        final String expectedCigar = "7M";
+        final SWPairwiseAlignment sw = new SWPairwiseAlignment(reference.getBytes(), read.getBytes(),
```
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchLeadingIgnoreLong\(\): String\.getBytes\(\)

sputnikci replied:
\[FindBugs\] WARNING: Dm: Found reliance on default encoding in org\.broadinstitute\.hellbender\.utils\.smithwaterman\.SWPairwiseAlignmentUnitTest\.testSubstringMatchLeadingIgnoreLong\(\): String\.getBytes\(\)

akiezun commented:
merging this
